### ISSUE TITLE
Proposal: Allow for 3rd party hooks into VHA

### ISSUE THREAD

**Issue by TheTacoScott** [whyboris/Video-Hub-App#477]
Since the code is open source, the custom hashing algorithm is easy to write in another language, the main config file (.vha2) is json, and the thumbnails/clips/filmstrips are live in predictable places with predictable file formats I think there's some benefit to offloading future "obscure" or "limiting" features to a simple 3rd party api that you expose from vha.

For example, if someone wants to regenerate all the clips using some custom nonsense they write (I'm speaking very much about myself here) or change the thumbnail of a clip every day using a custom thumbnail gen thing, this is all pretty doable by outputting different, but compliant, mp4/jpg at the correct location.

But we then need a way to send a signal to the app saying "things have change for hash X or video Y or whatever" if we want that update reflected at runtime.

An exposed API interface, or an ability to send and send messages to vha stdin or something similar to would open up the app.

Some of the other issues I've seen (like people who run vha on a separate server but use it on a different client) could be facilitated by opening up this framework.

TLDR: provide a way to send signals to a running VHA app (tcp port/rest api/stdin/etc) to let it know an external program has updated the state of:

* the json definition of a clip
* the thumbnail/clip/filmstrip of a clip

**Comment by TheTacoScott** [comment]
This piece fully offloads some obscure use cases out of the Video-Hub-App project and makes it so those features become extensions or plugins, or something similar. 

The project could facilitate their existence but isn't on the hook for their development.

**Comment by whyboris** [comment]
Thank you for the idea. I'm very unfamiliar with how a plug-in system / 3rd party hooks would work. The biggest challenge would be for me to understand what users could even want to do -- because unless I have a clear picture of at least some examples, I wouldn't be able to provide anything useful as an API.

In short, this would be a very large undertaking for me and since I wouldn't be using the feature, it's unlikely I'll be motivated to add it.

I am curious about why something like this is needed. The code is open-source -- anyone could edit the code however they like for a custom behavior, rather than use whatever not-far-sighted API that I could produce. I guess the benefit is that the API could be very small and well-documented, meaning no one has to look at the VHA code to get something added 🤔 

About updating the `.json` file while it's open in the app: I am unsure how it would work. For example, a user adds a tag to a video. Now the two data objects are different. If we don't allow external sources to update tags, we could make sure that VHA always does extra work to re-add the tags to incoming 'new' data ... but it seems super-complicated. It's just best not to have anything edit the .vha2 file when an app is open.

VHA 3.0.0 will allow the app to watch directories - adding new videos as they appear on the hard drive. So that solves at least one problem.

I understand that it would be cool to have a server that efficiently generates the screenshots (rather than doing it locally). But it seems easier to rip out the bit of code that does the extraction, write a script that keeps adding the screenshots without ever editing the .vha file. And when the app (running locally) imports the new videos, it finds the screenshots there (so it never tries to extract them). 🤷 

Do let me know if I've misunderstood -- and please feel free to add examples of what you imagine would be useful.

**Comment by whyboris** [comment]
I think I have a better idea of what you mean now that I think of it ... the `.vha2` (or `.json` as I referred to it above) wouldn't be updated by the external entity -- instead VHA would receive a message from the external entity about something changing somewhere. Is that right?

I have at best a very vague picture of how any of this would work (how could another language communicate to the app? would the plugins only work if they are written in a specific language?) Any time I can think of anything, the simplest solution is simply for whoever wants the functionality to edit the app themselves ... 

Please do clarify if you have the time -- because I never used a plugin in any of my software, so I'm likely missing out on some cool stuff!

**Comment by TheTacoScott** [comment]
Think even simpler than a plugin.

For the purposes of this issue let's just assume that a rest/json based web interface is what exists. 
Let's assume that vha listens on a configurable port, for this discussion let's assume port 9999, and listening on 0.0.0.0.

Let's limit to one use case: an image thumbnail (on disk) has been updated outside of the program and we want to reload it in the interface at runtime.

Here's how that could work:
POST http://<ip_of_vha>:9999/update

payload:
{
  "action": "thumbnail_update",
  "vha_hash": "<hash>",
}

That's not really very resty, but it could be refactored to something like

PUT http://<ip_of_vha>:9999/thumbnails/<vhash>

representing to vha to "update" that thumbnail for that vhash from disk

**Comment by TheTacoScott** [comment]
The benefit of something like this you don't need to know the exact use case, but you can expose a subset of the total feature set without having to worry about a way to make the gui work.

As of now the highest value targets would be a way to update the config file (vha2) and thumbnails/clips/filmstrips outside of the runtime and signal to the runtime that this has occurred.

**Comment by whyboris** [comment]
Oh yeah! A server with an HTTP endpoint 😅  -- makes sense

**Comment by TheTacoScott** [comment]
My usually style of markdown is getting clobbered here, but you get the idea. =)

**Comment by TheTacoScott** [comment]
https://github.com/mpc-hc/mpc-hc is a good example of a video player (much much simpler than vha) but it has a web interface that exposes... basically all the buttons in the gui to be "clicked" via web calls.

**Comment by TheTacoScott** [comment]
Now the benefit of this is: you don't really need to know the use cases. This is about letting others discovery what they can do. =)

Although specifically for me: I want the ability to signal to the app that a thumbnail has been updated, and that tags have been updated.

**Comment by whyboris** [comment]
FYI: currently you can:
- select any of the extracted screenshots (ones seen in the filmstrip) to be the default thumbnail per video
- you can drag/drop any image file over any video (in thumbnails view) and it will become the default thumbnail for that video

**Comment by TheTacoScott** [comment]
Figuring out a way to simulate a drag/drop was a path I went down, but sense this is electron and everything in the app isn't a classic windows "control" I abandoned that.

Picture a background process that is running, occasional;y updated thumbnails based on criteria not worth mentioning here. Having a way to have those updated without an up reload would be huge.

**Comment by whyboris** [comment]
😉  makes sense

I wouldn't want the app to automatically run a server in the background, but I could see this as a feature that's present (but not enabled in production builds -- meant more for anyone who wants to tinker on their own).

So, we could have a global toggle in the `node` section that is `false` by default -- never turning on the server, but if a developer wants to tinker, they enable it and we're ready to go.

The server would start and would listen for POST calls to some local address.

Upon the calls - it would simply use the `ipc` to send the messages to _Electron_ and the app would handle them however the developer wants to.

I think it may be simplest to have an example back-forth communication example with the non-enabled server and a bit of documentation in the `README.md` file to explain how to use it.

After that, it's up to every developer to put in whatever API they'd like.

In your case, you would just make a simple call to invalidate the image cache (already happening after an image is dropped over a thumbnail to replace it).

ps -- no need to use offset to change the thumbnail -- you look at the already-present functionality for picking any of the already-extracted thumbnails 👍

**Comment by whyboris** [comment]
@cal2195 -- I believe you've already played around with having some server stuff in _VHA_ in the past. If you're still interested in this topic -- please weigh in with some thoughts 🤝

**Comment by TheTacoScott** [comment]
I haven't done angular since angular 1, and have not touched nodejs since like 2015 but I will attempt to tackle this as well. 

If I dive into all these readmes and such I assume an otherwise reasonable person can get a dev env going?

**Comment by whyboris** [comment]
A slightly different direction, but FYI:
https://github.com/whyboris/Video-Hub-App/compare/master...cal2195:express-server
I think this plays videos over a locally-running server 🤷

**Comment by whyboris** [comment]
@TheTacoScott -- Angular has moved on since AngularJS -- so it may be easiest to treat it as if you're learning anew.

The good news is that once you `npm install` and `npm start` you'll have it all running.

Most of the functionality is in `home.component.ts` 
To invalidate the cache (and re-read all the images on disk) run this line:
https://github.com/whyboris/Video-Hub-App/blob/master/src/app/components/home.component.ts#L586
this.electronService.webFrame.clearCache();

I'm happy to help with any step of the way 🚀 🤝

**Comment by TheTacoScott** [comment]
whyboris: Why can't more people be like you? Love it. =)

**Comment by cal2195** [comment]
Hey guys! 😄 

This is certainly and interesting idea, and I think that an REST api is the way to go. As Boris mentioned, I have a working server on my fork for streaming (and even transcodeing) videos in-app, allowing for some cool extra features. 👍 

Expanding on this to allow access to more features, and controlling the app would be very cool - triggering a rescan remotely for example would be very cool! 😄 

One thing of note - whatever we choose should be planned out well and very extensible and flexible. If this is a dev feature, it mightn't matter as much, although I would hate to see custom code no longer work if we overhaul something because we missed some big issue! 💯 

I'm also interested in some of the other features @TheTacoScott has brought up in other issues, and I'll weigh in on my side on those as well!

I'm happy to help work on this, and design an extensible yet robust system that would serve many needs in the future! 👍

**Comment by TheTacoScott** [comment]
cal2195: Why can't more people be like you? Love it. =)

**Comment by TheTacoScott** [comment]
I think a web based resty/json like api makes a lot of senes (albeit I am very new to this project).

Proper modeling of the collections a resources to make it as discover-able and compliant as possible seems like a solid start after a POC is going

**Comment by whyboris** [comment]
I'm open to this "server on the side" approach. I would prefer to focus on getting VHA 3.0.0 released first before getting this code into the codebase. I'm hoping for a release in a month or two. Until then we could work on it, it just wouldn't become part of the `master` branch (soon to be renamed to `main` btw) until version 3 is out.

So the basic idea is a web server at some URL that is able to send messages to node which in turn can forward them to the 'front end' 🤝 

I'm happy to collaborate on this 👍

**Comment by cal2195** [comment]
I will dive in more deeply into this, and the other issues with ideas tomorrow morning - getting lateish here! 😄👍

**Comment by TheTacoScott** [comment]
So assuming that added a server side resty thing in nodejs is on the light end of trivial I'm curious what @cal2195 and @whyboris think of as the high level collections and resources for vha?

Forgive the newness (and the names can change to better match whatever) but I currently view high level collections as: **Hubs** and **Videos**

Since vha is single hub only (iirc), hubs may not make sense as a collection to expose.
Everything would likely be prefixed with a version ID

GET /hubs/
list hubs

GET /hubs/**HubName**
list hub configuration (inputDir/version/numOfFolders/etc)

GET /hubs/**HubName**/videos/
list videos (in this case the contents of vha.json["images"][*]

GET /hubs/**HubName**/videos/**hash**
vha.json["images"][**properindex**]

GET /hubs/**HubName**/videos/**hash**
vha.json["images"][**properindex**]

GET /hubs/**HubName**/videos/**hash**/(clip|thumbnail|filmstrip)
return an html compliant header and the image/video itself
Cache-Control: no-cache maybe could be used to invalidate cache here
This is a little non-resty since these clips and thumbnails and such are 1:1 and not 1:X relations but something like this would work.

PUT /hubs/**HubName**/videos/**hash**/(clip|thumbnail|filmstrip)
update the image/video with the payload data

DELETE /hubs/**HubName**/videos/**hash**/(clip|thumbnail|filmstrip)
delete the image/video, and maybe regenerated?


Not all of these need to be implemented, but this could be the foundation

Assuming something like this is sane, should we have a discussion with an OpenAPI spec or something similar on a PR?

**Comment by TheTacoScott** [comment]
Working POC: https://github.com/whyboris/Video-Hub-App/pull/478/files

**Comment by whyboris** [comment]
Related: #161 😊

**Comment by whyboris** [comment]
We're now one step closer as VHA has "official" server code 🎉 with #603
